hoomd/mpcd/DryingDropletStreamingMethod.cc

```diff
@@ -0,0 +1,125 @@
+
+/*!
+ * \param timestep Current time to stream
+ */
+void mpcd::DryingDropletStreamingMethod::stream(unsigned int timestep)
+    {
+    // use peekStream since shouldStream will be called by parent class
+    if(!peekStream(timestep)) return;
+
+    // compute final Radius and Velocity of surface
+    const Scalar start_R = m_R->getValue(timestep);
+    const Scalar end_R = m_R->getValue(timestep + m_period);
+    const Scalar V = (end_R - start_R)/(m_mpcd_dt);
+    // make sure that V < 0, since size of droplet must decrease
+    if (V > 0)
+        {
+        throw std::runtime_error("Droplet radius must decrease.");
+        }
+
+    /*
+     * If the initial geometry was not set, set the geometry and validate it.
+     * Because the interface is shrinking, it is sufficient to validate only the first time the geometry
+     * is set.
+     */
+    if (!m_geom)
+        {
+        m_geom = std::make_shared<mpcd::detail::SphereGeometry>(start_R, V, m_bc);
+        validate();
+        m_validate_geom = false;
+        }
+
+    /*
+     * Update the geometry radius to the size at the end of the streaming step.
+     * This needs to be done every time.
+     */
+    m_geom = std::make_shared<mpcd::detail::SphereGeometry>(end_R, V, m_bc);
+
+    // stream according to base class rules
+    ConfinedStreamingMethod<mpcd::detail::SphereGeometry>::stream(timestep);
+
+    // calculate number of particles to evaporate such that solvent density remains constant
+    const unsigned int N_end = std::round((4.*M_PI/3.)*end_R*end_R*end_R*m_density);
+    const unsigned int N_global = m_mpcd_pdata->getNGlobal();
+    const unsigned int N_evap = (N_end < N_global) ? N_global - N_end : 0;
+
+    /*
+     * Pick N_evap particles out of total number of bounced particles using RandomPicker.
+     * Npick is the total number of particles picked on this rank, m_picks will contain the indices of
+     * the picked particles.
+     */
+    unsigned int Npick = 0;
+    m_picker(m_picks, Npick, m_bounced, N_evap, timestep, m_mpcd_pdata->getN());
+
+    /*
+     * Apply the picks in the bounced indices by setting the next bit up.
+     */
+    const unsigned int mask = 1 << 1;
+    {
+    ArrayHandle<unsigned int> h_picks(m_picks, access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_bounced(m_bounced, access_location::host, access_mode::readwrite);
+    for (unsigned int i=0; i < Npick; ++i)
+        {
+        h_bounced.data[h_picks.data[i]] |= mask;
+        }
+    }
+
+    // finally, remove the picked particles using the bounced flags and mask
+    m_mpcd_pdata->removeParticles(m_removed,
+                                  m_bounced,
+                                  mask,
+                                  timestep);
+
+    // warn if density is too far from target
+    Scalar V_end = (4.*M_PI/3.)*end_R*end_R*end_R;
+    Scalar currentdensity = m_mpcd_pdata->getNGlobal()/V_end;
+    if (std::fabs(currentdensity - m_density) > Scalar(0.1))
```

Comment:
Should we make this a percentage rather than fixed amount? This would be 10%:
```suggestion
    if (std::fabs(currentdensity - m_density) > Scalar(0.1) * m_density)
```